### DISCUSSION TITLE
Sending Request Object by reference

### DISCUSSION THREAD

**Discussion by PrarthonaPaul** [keycloak/keycloak#23150] [Q&A] [answered]
Hello there, 
I am trying to add the ability to [specify the authentication request format](https://openid.net/specs/openid-connect-core-1_0.html#JWTRequests) for an application server. OpenID allows the client to send the Request object as a jwt that can be signed and optionally encrypted. 

Now, this can be sent by value, using the `request` [parameter](https://openid.net/specs/openid-connect-core-1_0.html#RequestObject) or by reference using the `request_uri` [parameter](https://openid.net/specs/openid-connect-core-1_0.html#RequestUriParameter). In the second case, the openID server accesses the endpoint specified by request-uri to access the jwt. 

Now my question is, does Keycloak support sending request parameters by reference using `request_uri`? if no, how can the client generate this `request-uri` once we have converted the Request Object to a jwt? 

And my second question is about encrypting. So far, I have been able to figure out how to send the request object as a signed jwt. However, when I encrypt the RSA signed jwt using the same KeyPair, Keycloak is unable to decrypt it. Here is what the application and console messages are: 
![image](https://github.com/keycloak/keycloak/assets/83880425/39e168c7-ab6e-40ea-98c6-0cb067be5147)
keycloak console message: 
```
WARN  [org.keycloak.services] (executor-thread-83) KC-SERVICES0097: Invalid request: java.lang.RuntimeException: Failed to decrypt JWT
	at org.keycloak.jose.jws.DefaultTokenManager.decodeClientJWT(DefaultTokenManager.java:162)
	at org.keycloak.protocol.oidc.endpoints.request.AuthzEndpointRequestObjectParser.<init>(AuthzEndpointRequestObjectParser.java:45)
	at org.keycloak.protocol.oidc.endpoints.request.AuthorizationEndpointRequestParserProcessor.parseRequest(AuthorizationEndpointRequestParserProcessor.java:88) 
```
followed by: 
```
 WARN  [org.keycloak.events] (executor-thread-104) type=LOGIN_ERROR, realmId=91573453-cb09-4c53-b022-8b98255d6b3a, clientId=myclient, userId=null, ipAddress=172.17.0.1, error=invalid_request
WARN  [org.keycloak.events] (executor-thread-116) type=REFRESH_TOKEN_ERROR, realmId=03997366-1a2a-47e8-b42b-dd805b24ca69, clientId=security-admin-console, userId=null, ipAddress=172.17.0.1, error=invalid_token, grant_type=refresh_token, client_auth_method=client-secret
```
I am using the same keyPair to sign and encrypt the jwt and following the specs to sign first and then encrypt to make a nested jwt. But keycloak is having no trouble reading the signature, but cannot seem to decrypt it. 
 I am not sure what is happening. Any help would be appreciated! 
Thank you!

**Comment by justin-tay** [comment, marked as answer]
You can either use the PAR endpoint to get the `request_uri` from Keycloak or expose a uri on your application for Keycloak to retrieve from. 

For PAR you need to send your request first to the `pushed_authorization_request_endpoint` eg. http://keycloak/realms/{realm}/protocol/openid-connect/ext/par/request. The endpoint will return the `request_uri`. The `request_uri` here is something Keycloak generates to identify the request.

When specifying a `request_uri` on your application for Keycloak to retrieve from you need to ensure that the `request_uri` specified is configured in `Valid request URIs` in your client's `Fine grain OpenID Connect configuration`.

![requesturi](https://github.com/keycloak/keycloak/assets/49700559/56c7266b-7cf4-4b58-ab87-abdaf2164a94)

You need to use Keycloak's public encryption keys if you want to encrypt something for Keycloak. This can be obtained from the `jwks_url` eg. http://keycloak/realms/{realm}/protocol/openid-connect/certs. The keys with `use` `enc` is the ones you should use. You can also see these keys in `Realm settings`.

![realmkeys](https://github.com/keycloak/keycloak/assets/49700559/b2d26905-90ca-4e76-81d3-91ebde34aa3a)

**Reply by PrarthonaPaul** [reply]
Thank you! I used the PAR method and it works! And using RSA_OAEP public key to encrypt jwt's works perfectly!
